Chore: Use AYON entities

iLLiCiTiT opened the pull request:
## Changelog Description
This PR is removing `ayon_core.client` and replaces it's usage with `ayon_api` calls. That changed structure of all entity types from OpenPype compatible to AYON.

## Additional info
I did what I could to change every line of code which are working with entities to work properly, but I've probably missed something so please report anything you hit. And I didn't test most of the changes.

Entity structure changes sometimes required change of logic and function arguments. Some of those changes happened on the spot, so if you hit a change that is questionable then please comment. I'll mention few questionable changes myself. In some cases I did more changes than just change of query and used keys because it helper me with other changes, or I couldn't resist to refactor few lines.

## Major changes
- Project, asset, subset, version, representation and workfile documents are now project, folder, product, version, representation and workfile entities.
- Removed `ayon_core.client`. Its functions were replaced with function from `ayon_api`.
- Implemented `initialize_ayon_connection` in `ayon_core.lib.ayon_connection` to initialize ayon connection. Initialization of the connection happened automatically when a call in `ayon_core.client` happened. It is automatically called when `cli.main` is called and on host install or ayon plugin install, so all important places should be covered. This affects local root overrides so it is important to happen as soon as possible in any process that might work with Anatomy.
- Added usage of task entity. Before this PR task information was on asset document, but AYON has task entity separated from folder entity so many places did have to change. Almost all places did expect asset document and task name and it was enough, now is usually needed only task entity, which affected arguments of functions (e.g. `get_product_name`).
- Renamed multiple funtions which contained asset, e.g. `get_current_asset_name` > `get_current_folder_path`, `get_current_project_task` > `get_current_project_folder`.
- Context in load plugins does not have `"asset"` and `"subset"` keys, but `"folder"` and `"product"`.

### Notes
I was very struggling with changes in unreal host. There is a lot of "asset" naming that is not related to OpenPype asset, and also there was a lot of unusuall way of working with representations and what is stored to container data. If it is working as expected now then it's fine, details can be changed when unreal will be in separated addon.

### Questions
- ~~Function `get_product_name` now expects task entity, but from api point of view it might be better to expect `task_name` and `task_type` as that's what is really needed for the function.~~ Changed the function to expect task name and task type instead of task entity.
- Function `initialize_ayon_connection` is really important and I'm not sure if should stay in `ayon_core.lib`, I couldn't find more suitable place and name.

## Testing notes:
Please consider that some of the bugs may not come from this PR, as was found out many times in previous PRs.

Keep in mind that this PR will completelly break compatibility with almost any other addon, so ftrack, shotgrid, kitsu, slack etc. won't work.

Test every host, every addon, every possible action that can be done should be tested. Create, Publish, Load, Manage in scene inventory. Open old scene, with previously created instances and loaded containers, all of that should be working as nothing changed (the scenes should be from AYON mode).

All tools should show correct data and should work as expected. Console should not log any errors.
- Publisher allows to create all kind of instances, change their context and product name, change attributes. Crashed plugins list does not contain anything suspicious (usually there are deadline plugins). Publishing works. Integration creates entities and copies files. Thumbnails are integrated etc.
- Workfiles tool should open, save as both workfiles and published workfiles. Show notes about the workfile etc.
- Loader tool should load.
- Scene inventory can change version, update to latest, or change folder/product/representation.

### Hosts
- [x] 3dsMax
- [x] Aftereffect
- [x] Blender
- [x] Fusion
- [x] Hiero
- [x] Houdini
- [x] Maya
- [x] Nuke
- [x] Photoshop
- [ ] Resolve - almost done
- [x] SubstancePainter
- [x] TVpaint
- [x] Unreal

### Other
- [ ] Push to project works

kalisp commented on client/ayon_core/plugins/publish/collect_scene_loaded_versions.py:
```diff
@@ -42,21 +42,21 @@ def process(self, context):
             repre_id = con["representation"]
-            repre_doc = repre_doc_by_str_id.get(repre_id)
+            repre_entity = repre_entities_by_id.get(repre_id)
             if repre_doc is None:
```
```suggestion
            if repre_entity is None:
```

kalisp commented on client/ayon_core/plugins/publish/collect_anatomy_instance_data.py:
```diff
@@ -51,73 +51,173 @@ def process(self, context):
+        context_task_entity = context.data.get("taskEntity")
+        context_task_name = None
+        if context_task_entity:
+            context_task_name = context_task_entity["path"]
```
```suggestion
            context_task_name = context_task_entity["name"]
```

moonyuet commented:
When testing with 3dsMax, it errors out in integrate hero version. 
```
DEBUG: --- Integration of Hero version for product `modelMain` begins.
DEBUG: Looking for matching profile for: hosts: "max" | product_types: "model" | task_names: "art" | task_types: "Art"
DEBUG: "max" not found in "hosts": ['standalonepublisher']
DEBUG: None of profiles match your setup. hosts: "max" | product_types: "model" | task_names: "art" | task_types: "Art"
DEBUG: `hero` template check was successful. `{root[work]}/{project[name]}/{hierarchy}/{folder[name]}/publish/{product[type]}/{product[name]}/hero/{project[code]}_{folder[name]}_{task[name]}_hero<_{comment}>.{ext}`
DEBUG: hero publish dir: "D:\sh_boss_proj\interior_lenny\lenny_bedroom\asset\publish\model\modelMain\hero"
DEBUG: Response <RestApiResponse [200]>
DEBUG: Response <RestApiResponse [200]>
DEBUG: Replacing old hero version.
DEBUG: Backup folder path is "D:\sh_boss_proj\interior_lenny\lenny_bedroom\asset\publish\model\modelMain\hero.BACKUP"
ERROR: !!! Creating of hero version failed. Previous hero version maybe lost some data!
Traceback (most recent call last):
  File "C:\Users\Kayla\AppData\Local\Ynput\AYON\dependency_packages\ayon_2402141620_windows.zip\dependencies\pyblish\plugin.py", line 527, in __explicit_process
    runner(*args)
  File "D:\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 125, in process
    self.integrate_instance(
  File "D:\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 400, in integrate_instance
    repre_files = self.get_files_info(dst_paths, anatomy)
  File "D:\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 503, in get_files_info
    file_info = self.prepare_file_info(filepath, anatomy)
  File "D:\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 519, in prepare_file_info
    "id": create_entity_id(),
NameError: name 'create_entity_id' is not defined
```
The attachment is the publish report.
[publish-report-240311-19-18.json](https://github.com/ynput/ayon-core/files/14557699/publish-report-240311-19-18.json)

kalisp commented:
In Fusion now in integrate:
![image](https://github.com/ynput/ayon-core/assets/4457962/72d260a3-e78b-46af-83ac-b9e866b4aec2)

iLLiCiTiT commented:
> When testing with 3dsMax, it errors out in integrate hero version.

Fixed.

> In Fusion now in integrate:

Fixed.

MustafaJafar commented on client/ayon_core/hosts/houdini/api/lib.py:
```diff
@@ -827,41 +839,47 @@ def get_camera_from_container(container):
+    anatomy = Anatomy(project_name, project_entity=project_entity)
+    folder_entity = ayon_api.get_folder_by_path(project_name, folder_path)
+    task_entity = ayon_api.get_task_by_name(
+        project_name, folder_path["id"], task_name
```
There's a typo here.
```suggestion
        project_name, folder_entity["id"], task_name
```

MustafaJafar commented:
I was testing Farm submission and Houdini job was submitted fine. But, the ayon publish job didn't. 
I tried updating my env and dependency package but didn't work.
```
DEBUG: http://192.168.1.4:5000 "POST /graphql HTTP/1.1" 200 151
DEBUG: Response <RestApiResponse [200]>
Traceback (most recent call last):
  File "C:\Users\Mustafa Taher\AppData\Local\Ynput\AYON\dependency_packages\ayon_2403111503_windows.zip\dependencies\pyblish\plugin.py", line 527, in __explicit_process
    runner(*args)
  File "C:\Users\Mustafa Taher\AppData\Local\Ynput\AYON\addons\core_0.3.0-PR165.1\ayon_core\modules\deadline\plugins\publish\submit_publish_job.py", line 468, in process
    self._submit_deadline_post_job(instance, render_job, instances)
  File "C:\Users\Mustafa Taher\AppData\Local\Ynput\AYON\addons\core_0.3.0-PR165.1\ayon_core\modules\deadline\plugins\publish\submit_publish_job.py", line 187, in _submit_deadline_post_job
    output_dir = self._get_publish_folder(
  File "C:\Users\Mustafa Taher\AppData\Local\Ynput\AYON\addons\core_0.3.0-PR165.1\ayon_core\modules\deadline\plugins\publish\submit_publish_job.py", line 538, in _get_publish_folder
    version_entity = ayon_api.get_last_version_by_product_name(
  File "E:\Ynput\ayon-launcher\.venv\lib\site-packages\ayon_api\_api.py", line 895, in get_last_version_by_product_name
    return con.get_last_version_by_product_name(*args, **kwargs)
  File "E:\Ynput\ayon-launcher\.venv\lib\site-packages\ayon_api\server_api.py", line 4866, in get_last_version_by_product_name
    product = self.get_product_by_name(
  File "E:\Ynput\ayon-launcher\.venv\lib\site-packages\ayon_api\server_api.py", line 4340, in get_product_by_name
    for product in products:
  File "E:\Ynput\ayon-launcher\.venv\lib\site-packages\ayon_api\server_api.py", line 4248, in get_products
    parsed_data = query.query(self)
  File "E:\Ynput\ayon-launcher\.venv\lib\site-packages\ayon_api\graphql.py", line 343, in query
    raise GraphQlQueryFailed(response.errors, query_str, variables)
ayon_api.exceptions.GraphQlQueryFailed: GraphQl query Failed: Cannot query field '_id' on type 'ProductNode'. Did you mean 'id'? (Line 6 Column 11)
```

moonyuet commented:
another error pops up when speaking of testing the branch in 3dsMax.
```
DEBUG: --- Integration of Hero version for product `modelMain` begins.
DEBUG: Looking for matching profile for: hosts: "max" | product_types: "model" | task_names: "art" | task_types: "Art"
DEBUG: "max" not found in "hosts": ['standalonepublisher']
DEBUG: None of profiles match your setup. hosts: "max" | product_types: "model" | task_names: "art" | task_types: "Art"
DEBUG: `hero` template check was successful. `{root[work]}/{project[name]}/{hierarchy}/{folder[name]}/publish/{product[type]}/{product[name]}/hero/{project[code]}_{folder[name]}_{task[name]}_hero<_{comment}>.{ext}`
DEBUG: hero publish dir: "D:\sh_boss_proj\interior_lenny\lenny_bedroom\asset\publish\model\modelMain\hero"
DEBUG: Response <RestApiResponse [200]>
DEBUG: Response <RestApiResponse [200]>
DEBUG: Replacing old hero version.
DEBUG: Backup folder path is "D:\sh_boss_proj\interior_lenny\lenny_bedroom\asset\publish\model\modelMain\hero.BACKUP"
ERROR: !!! Creating of hero version failed. Previous hero version maybe lost some data!
Traceback (most recent call last):
  File "C:\Users\Kayla\AppData\Local\Ynput\AYON\dependency_packages\ayon_2402141620_windows.zip\dependencies\pyblish\plugin.py", line 527, in __explicit_process
    runner(*args)
  File "D:\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 126, in process
    self.integrate_instance(
  File "D:\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 401, in integrate_instance
    repre_files = self.get_files_info(dst_paths, anatomy)
  File "D:\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 504, in get_files_info
    file_info = self.prepare_file_info(filepath, anatomy)
  File "D:\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 522, in prepare_file_info
    "path": self.get_rootless_path(anatomy, path),
AttributeError: 'IntegrateHeroVersion' object has no attribute 'get_rootless_path'

```

kalisp reviewed (approved):
Tested in PS, AE, Fusion

moonyuet commented:
Tested in Substance Painter, Looks good.
Will approve after the error in Max has been resolved.

iLLiCiTiT commented:
> I tried updating my env and dependency package but didn't work.

Damn, this is bug in ayon api.

jakubjezek001 commented:
> Keep in mind that this PR will completelly break compatibility with almost any other addon, so ftrack, shotgrid, kitsu, slack etc. won't work.

Is this really not possible to make backward compatible changes so it is not breaking addon compatibility?

moonyuet commented:
Testing in Max and face this
```
DEBUG: --- Integration of Hero version for product `modelMain` begins.
DEBUG: Looking for matching profile for: hosts: "max" | product_types: "model" | task_names: "art" | task_types: "Art"
DEBUG: "max" not found in "hosts": ['standalonepublisher']
DEBUG: None of profiles match your setup. hosts: "max" | product_types: "model" | task_names: "art" | task_types: "Art"
DEBUG: `hero` template check was successful. `{root[work]}/{project[name]}/{hierarchy}/{folder[name]}/publish/{product[type]}/{product[name]}/hero/{project[code]}_{folder[name]}_{task[name]}_hero<_{comment}>.{ext}`
DEBUG: hero publish dir: "D:\sh_boss_proj\interior_lenny\lenny_bedroom\asset\publish\model\modelMain\hero"
DEBUG: Response <RestApiResponse [200]>
DEBUG: Response <RestApiResponse [200]>
DEBUG: Replacing old hero version.
DEBUG: Backup folder path is "D:\sh_boss_proj\interior_lenny\lenny_bedroom\asset\publish\model\modelMain\hero.BACKUP"
DEBUG: D:/sh_boss_proj/interior_lenny/lenny_bedroom/asset/publish/model/modelMain/hero/il_asset_art_hero.abc
ERROR: !!! Creating of hero version failed. Previous hero version maybe lost some data!
Traceback (most recent call last):
  File "C:\Users\Kayla\AppData\Local\Ynput\AYON\dependency_packages\ayon_2402141620_windows.zip\dependencies\pyblish\plugin.py", line 527, in __explicit_process
    runner(*args)
  File "D:\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 126, in process
    self.integrate_instance(
  File "D:\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 401, in integrate_instance
    repre_files = self.get_files_info(dst_paths, anatomy)
  File "D:\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 504, in get_files_info
    file_info = self.prepare_file_info(filepath, anatomy)
  File "D:\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 524, in prepare_file_info
    "size": os.path.getsize(path),
  File "C:\Program Files\Autodesk\3ds Max 2024\Python\lib\genericpath.py", line 50, in getsize
    return os.stat(filename).st_size
FileNotFoundError: [WinError 3] The system cannot find the path specified: 'D:/sh_boss_proj/interior_lenny/lenny_bedroom/asset/publish/model/modelMain/hero/il_asset_art_hero.abc'
```

iLLiCiTiT commented:
> Is this really not possible to make backward compatible changes so it is not breaking addon compatibility?

No.

jakubjezek001 commented:
> > Is this really not possible to make backward compatible changes so it is not breaking addon compatibility?
> 
> No.

Your short answer `No` does say nothing about why it is not possible. Could you please explain why is it not possible since not only me might be interested about it. Please make it ideally more then one sentence, thank you.

jakubjezek001 reviewed (changes requested):
Publishing from Nuke on farm is broken?

 ERROR:pyblish.plugin:Traceback (most recent call last):
2024-03-11 15:48:29:  0: STDOUT:   File "C:\Users\jakubjezek001\AppData\Local\Ynput\AYON\dependency_packages\ayon_2401161802_windows.zip\dependencies\pyblish\plugin.py", line 527, in __explicit_process
2024-03-11 15:48:29:  0: STDOUT:     runner(*args)
2024-03-11 15:48:29:  0: STDOUT:   File "C:\CODE\__PYPE\ayon-core\client\ayon_core\plugins\publish\collect_anatomy_instance_data.py", line 54, in process
2024-03-11 15:48:29:  0: STDOUT:   File "C:\CODE\__PYPE\ayon-core\client\ayon_core\plugins\publish\collect_anatomy_instance_data.py", line 73, in fill_missing_folder_entities
2024-03-11 15:48:29:  0: STDOUT: KeyError: 'task'
2024-03-11 15:48:29:  0: STDOUT: Traceback (most recent call last):
2024-03-11 15:48:29:  0: STDOUT:   File "C:\Users\jakubjezek001\AppData\Local\Ynput\AYON\dependency_packages\ayon_2401161802_windows.zip\dependencies\pyblish\plugin.py", line 527, in __explicit_process
2024-03-11 15:48:29:  0: STDOUT:     runner(*args)
2024-03-11 15:48:29:  0: STDOUT:   File "<string>", line 54, in process
2024-03-11 15:48:29:  0: STDOUT:   File "<string>", line 73, in fill_missing_folder_entities
2024-03-11 15:48:29:  0: STDOUT: KeyError: 'task'
2024-03-11 15:48:29:  0: STDOUT: !!! ERR: 2024-03-11 15:48:28,904 >>> { CLI-publish }: [ Failed CollectAnatomyInstanceData: 'task' -- ('C:\\CODE\\__PYPE\\ayon-core\\client\\ayon_core\\plugins\\publish\\collect_anatomy_instance_data.py', 73, 'fill_missing_folder_entities', '') ]

BigRoy commented:
> > Your short answer No does not say anything about why it is not possible. Could you please explain why is it not possible since not only me might be interested about it. Please make it ideally more then one sentence, thank you.
> 
> The data structure of all entities that are used in almost every pipeline function, plugin and metadata did change completelly. I don't think there is anything. It would be like trying to add support to all functions to be able to use both ftrack entities and shotgrid entities.
> 
> In other words it might be maybe possible, but the time would be in months of work instead of days and the effort would not worth.

Would technically keeping `.client` around with deprecation warnings make it backwards compatible enough for the addons to continue working and move over in their own pace?

iLLiCiTiT commented:
> Would technically keeping `ayon_core.client` around with deprecation warnings make it backwards compatible enough for the addons to continue working and move over in their own pace?

What should the functions return, OpenPype-like documents or AYON entities? If OpenPype-like entities, then you can't use them in any available functions, if AYON entities you can't work with them as you would before this PR.

Example:
```python
from ayon_core.client import get_representations
from ayon_core.pipeline import get_representation_path


repres = list(
    get_representations("someProject", representation_names={"look"})
)

# This would crash with AYON-like entities
repre_ids = {repre["_id"] for repre in repres}

# This would crash with OpenPype-like documents
repre_paths = {
    get_representation_path(repre)
    for repre in repres
}
```

BigRoy commented:
Ah I see, yes. This is a major refactor then - I figured the returns would at least be 'similar' in data. Is there a reference of 'what's changed'? By the way, I had no idea `.client` was intended to be removed only now I realize the amount of code I have that calls `.client`. I'd love an overview of changes or "what is the equivalent api call" etc. reference.

Is it just the lack of idea? Is the full return type different? (Is it still a `dict`?) Is `"data"` still a thing in the entities or is it structured differently? If so, is there some document or reference regarding the new data structure?

Even though we might be refactoring much more things "soon" would it make sense to version bump all addons with e.g. a minor or even major version with such drastic changes. Yes, we might be at `core-5.0.0` rather quickly but at least we have a simple major version identifier of what things were broken at what major release. It may also make sense to similarly say that `0.0.15` maya addon or whatever is incompatible with `core-1.0.0` and you need to have `maya-1.0.0` at least or accordingly. Of course this is very rudimentary but at least might be some poor man's compatibility check someone can though based on solely the version numbers of the addon?

jakubjezek001 commented:
I was testing the Deadline publishing from Nuke again. Still fails but this time on different point.

```
2024-03-11 21:42:50:  0: STDOUT: ERROR:pyblish.plugin:Traceback (most recent call last):
2024-03-11 21:42:50:  0: STDOUT:   File "C:\Users\jakubjezek001\AppData\Local\Ynput\AYON\dependency_packages\ayon_2401161802_windows.zip\dependencies\pyblish\plugin.py", line 527, in __explicit_process
2024-03-11 21:42:50:  0: STDOUT:     runner(*args)
2024-03-11 21:42:50:  0: STDOUT:   File "C:\CODE\__PYPE\ayon-core\client\ayon_core\plugins\publish\collect_anatomy_instance_data.py", line 55, in process
2024-03-11 21:42:50:  0: STDOUT:   File "C:\CODE\__PYPE\ayon-core\client\ayon_core\plugins\publish\collect_anatomy_instance_data.py", line 209, in fill_missing_task_entities
2024-03-11 21:42:50:  0: STDOUT: TypeError: sequence item 1: expected str instance, NoneType found
2024-03-11 21:42:50:  0: STDOUT: Traceback (most recent call last):
2024-03-11 21:42:50:  0: STDOUT:   File "C:\Users\jakubjezek001\AppData\Local\Ynput\AYON\dependency_packages\ayon_2401161802_windows.zip\dependencies\pyblish\plugin.py", line 527, in __explicit_process
2024-03-11 21:42:50:  0: STDOUT:     runner(*args)
2024-03-11 21:42:50:  0: STDOUT:   File "<string>", line 55, in process
2024-03-11 21:42:50:  0: STDOUT:   File "<string>", line 209, in fill_missing_task_entities
2024-03-11 21:42:50:  0: STDOUT: TypeError: sequence item 1: expected str instance, NoneType found
```

iLLiCiTiT commented:
> Is it just the lack of _id? Is the full return type different? (Is it still a dict?) Is "data" still a thing in the entities or is it structured differently? If so, is there some document or reference regarding the new data structure?

It differs based on entity type. Key `"_id"` is first obvious. Key `"data"` is still there, but there is also `"attrib"`. Attributes are defined and some are built-in, but the `data` may still be used for "custom" data. Some keys moved from data elsewhere or into new entity. E.g. task is not on folder but is separate entity. Folder has `folderType`, task has `taskType`, product has `productType`, version does has `"version"` and `"name"` where version is integer, which is negative for hero versions and name is just formatted string `"v001"`. Key `"parent"` is not available either, it is `"folderId"`, `"taskId"`, `"productId"`, `"versionId"`, `"parentId"` for folders.

> Even though we might be refactoring much more things "soon" would it make sense to version bump all addons with e.g. a minor or even major version with such drastic changes. Yes, we might be at core-5.0.0 rather quickly but at least we have a simple major version identifier of what things were broken at what major release. It may also make sense to similarly say that 0.0.15 maya addon or whatever is incompatible with core-1.0.0 and you need to have maya-1.0.0 at least or accordingly. Of course this is very rudimentary but at least might be some poor man's compatibility check someone can though based on solely the version numbers of the addon?

All addons that are currently in ayon-core does not have to bump anything, their code lives in ayon-core... That's why the separation of addons should happen after all the refactoring. The ayon-core with this change will probably be `0.3.0` release. For all other addons (ftrack, shotgrid, etc.) it is a minor version bump.

LiborBatek commented:
```Blender``` seems working fine... did overall testing of all ```AYON``` tools and didnt find any issues so far...will continue with more elaborate testing but so far good!

simonebarbieri commented:
Unreal seems to be working fine!

iLLiCiTiT commented:
> I was testing Farm submission and Houdini job was submitted fine. But, the ayon publish job didn't.
I tried updating my env and dependency package but didn't work.

This is probably the worse bug, that is even not in this repository. Fixed by monkey patching ayon api.

> Testing in Max and face this

Fixed.

> Publishing from Nuke on farm is broken?

I think it's fixed.

> I am getting this error during the thumbnail integration in Blender:

Fixed.

iLLiCiTiT commented:
> I got this error in Houdini

Fixed.

> Another Error with Integrate Hero Asset.

Please try again. I think it's fixed, but.

iLLiCiTiT commented:
> when inspecting the report there is obviously missing Review product - also not just visually in the UI but its not published at all

Is that related to this PR, or it happens in develop too?

moonyuet reviewed (approved):
Works in Max now.

LiborBatek commented:
> > when inspecting the report there is obviously missing Review product - also not just visually in the UI but its not published at all
> 
> Is that related to this PR, or it happens in develop too?

It also does happen in the develop so not related to this PR

BigRoy commented:
> > > when inspecting the report there is obviously missing Review product - also not just visually in the UI but its not published at all
> > 
> > 
> > Is that related to this PR, or it happens in develop too?
> 
> It also does happen in the develop so not related to this PR

Yup, likely has to do with the review instance 'attaching' itself to the model main instance in this case so likely "removing" itself and merging the instance with the model main instance or alike.

MustafaJafar commented:
Cool, I no longer see previous errors. 
New Error 🎉, AYON publish job fails on farm.
```
2024-03-12 16:49:05:  0: STDOUT: ERROR:pyblish.plugin:Traceback (most recent call last):
2024-03-12 16:49:05:  0: STDOUT:   File "C:\Users\Mustafa Taher\AppData\Local\Ynput\AYON\dependency_packages\ayon_2403111503_windows.zip\dependencies\pyblish\plugin.py", line 527, in __explicit_process
2024-03-12 16:49:05:  0: STDOUT:     runner(*args)
2024-03-12 16:49:05:  0: STDOUT:   File "E:\Ynput\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 126, in process
2024-03-12 16:49:05:  0: STDOUT:   File "E:\Ynput\ayon-core\client\ayon_core\plugins\publish\integrate_hero_version.py", line 242, in integrate_instance
2024-03-12 16:49:05:  0: STDOUT: KeyError: 'taskId'
2024-03-12 16:49:05:  0: STDOUT: Traceback (most recent call last):
2024-03-12 16:49:05:  0: STDOUT:   File "C:\Users\Mustafa Taher\AppData\Local\Ynput\AYON\dependency_packages\ayon_2403111503_windows.zip\dependencies\pyblish\plugin.py", line 527, in __explicit_process
2024-03-12 16:49:05:  0: STDOUT:     runner(*args)
2024-03-12 16:49:05:  0: STDOUT:   File "<string>", line 126, in process
2024-03-12 16:49:05:  0: STDOUT:   File "<string>", line 242, in integrate_instance
2024-03-12 16:49:05:  0: STDOUT: KeyError: 'taskId'
2024-03-12 16:49:05:  0: STDOUT: !!! ERR: 2024-03-12 16:49:05,056 >>> { CLI-publish }: [ Failed IntegrateHeroVersion: 'taskId' -- ('E:\\Ynput\\ayon-core\\client\\ayon_core\\plugins\\publish\\integrate_hero_version.py', 242, 'integrate_instance', '') ] 
```

LiborBatek commented:
Didnt find any issues in ```Maya``` atm.... so far good too.

MustafaJafar commented:
Cool, I no longer see previous errors.
New Error 🎉, when clicking `Switch to versioned`
```
Traceback (most recent call last):
  File "E:\Ynput\ayon-core\client\ayon_core\tools\sceneinventory\view.py", line 234, in <lambda>
    lambda: _on_switch_to_versioned(items)
  File "E:\Ynput\ayon-core\client\ayon_core\tools\sceneinventory\view.py", line 187, in _on_switch_to_versioned
    standard_versions = ayon_api.get_versions(
  File "E:\Ynput\ayon-launcher\.venv\lib\site-packages\ayon_api\_api.py", line 855, in get_versions
    return con.get_versions(*args, **kwargs)
TypeError: get_versions() missing 1 required positional argument: 'project_name'
```

MustafaJafar commented:
So far in Houdini, So far so good. 
- [x] Create
- [x] Local Publish
- [x] Farm Publish 
- [x] Load
- [x] Manage

Regards workfiles: I hit an issue when switch context using workfiles tools.. And, I was able to reproduce the same behavior in develop.

iLLiCiTiT commented:
> I hit an issue when switch context using workfiles tools.. And, I was able to reproduce the same behavior in develop.

I'll dm you, but won't be fixed in this PR if happens in develop too. Please check `Houdini` checkbox if you finished with houdini testing, and everything seems to be working 🙂 .

jakubjezek001 commented:
Nuke is now working also with Farm rendering and publishing

jakubjezek001 commented:
In Resolve following error occurs if I try to start Creator from Ayon menu:

```
Traceback (most recent call last):
  File "C:\CODE\__PYPE\ayon-core\client\ayon_core\hosts\resolve\api\menu.py", line 138, in on_create_clicked
    host_tools.show_creator()
  File "C:\CODE\__PYPE\ayon-core\client\ayon_core\tools\utils\host_tools.py", line 383, in show_creator
    _SingletonPoint.show_tool_by_name("creator", parent)
  File "C:\CODE\__PYPE\ayon-core\client\ayon_core\tools\utils\host_tools.py", line 349, in show_tool_by_name
    cls.helper.show_tool_by_name(tool_name, parent, *args, **kwargs)
  File "C:\CODE\__PYPE\ayon-core\client\ayon_core\tools\utils\host_tools.py", line 309, in show_tool_by_name
    self.show_creator(parent, *args, **kwargs)
  File "C:\CODE\__PYPE\ayon-core\client\ayon_core\tools\utils\host_tools.py", line 113, in show_creator
    creator_tool.refresh()
  File "C:\CODE\__PYPE\ayon-core\client\ayon_core\tools\creator\window.py", line 384, in refresh
    get_current_project_settings()
KeyError: 'global'
```

It seems to be unrelated from this PR subject.

jakubjezek001 commented:
In Resolve during loading this had been blocking the loading. Again, I am not sure if it is related to this PRs changes.

```
Traceback (most recent call last):
  File "C:\CODE\__PYPE\ayon-core\client\ayon_core\tools\loader\models\actions.py", line 740, in _load_representations_by_loader
    load_with_repre_context(
  File "C:\CODE\__PYPE\ayon-core\client\ayon_core\pipeline\load\utils.py", line 323, in load_with_repre_context
    return loader.load(repre_context, name, namespace, options)
  File "C:\CODE\__PYPE\ayon-core\client\ayon_core\hosts\resolve\plugins\load\load_clip.py", line 45, in load
    timeline_item = plugin.ClipLoader(
  File "C:\CODE\__PYPE\ayon-core\client\ayon_core\hosts\resolve\api\plugin.py", line 415, in load
    if "slate" in self.data["versionData"]["families"]:
KeyError: 'versionData'
```

iLLiCiTiT commented:
> It seems that attrib is some key which suppose to be on version, since I was checking it and it occurs to be even in nuke, maya, vusion etc. But here it is missing.

Should be fixed, please revalidate.
